### PR TITLE
Add unit-wise for linear

### PR DESCRIPTION
**Pull request by Cecilwang**
- [x] linear
- [x] conv
- [x] unit test
- [x] unit-wise precondition
- [x] ~~training example~~

### CODE REVIEW

**Comment by Cecilwang** [comment]
placeholder

File: asdfghjkl/operations/linear.py [hunk 1]
```diff
@@ -44,6 +44,14 @@ def cov_kron_A(module, in_data):
...
         return torch.matmul(out_grads.T, out_grads)  # f_out x f_out
 
+    @staticmethod
+    def cov_unit_wise(module, in_data, out_grads):
+        if module.bias is not None:
+            in_data = torch.cat([in_data, torch.ones(in_data.size(0), 1)], 1)
+        in_in = torch.matmul(in_data.T, in_data)  # f_in x f_in
+        grad_grad = out_grads.mul(out_grads).sum(dim=0)  # f_out
+        return torch.einsum('i, jk -> ijk', grad_grad, in_in).to(in_data.device)
+
```
Comment (kazukiosawa) [inline asdfghjkl/operations/linear.py:54]: This logic seems not correct.

This is the unit-wise block-diagonal Fisher for the l-th layer

image

And each diagonal block is defined like this:

image

For now, just ignore `\mathbb{E}`. 
`M_{l-1}` is the input dimension of the l-th layer.
`M_l` is the output dimension of the l-th layer.
`<>` represents a data average.
`h` is the layer input (i.e., `in_data`)
`\nabla_u \ell` is the gradient w.r.t. layer output (i.e., `out_grads`) 

as you can see, `in_data` and `out_grads` "for each example" need to be multiplied by each other before taking an data average.

File: asdfghjkl/operations/operation.py [hunk 2]
```diff
@@ -26,6 +26,7 @@ def __init__(self, module, model, op_names, save_field='op_results'):
         self._save_field = save_field
         self._grads_scale = None
         self._kron_A = None
+        self._unit_wise_A = None
```
[the remark this answers is not in the document]
Comment (Cecilwang) [reply asdfghjkl/operations/operation.py:29]: This field is used to store extended input data. 
This data is calculated in `forward_post_process` and is used in `backward_pre_process`.

One way to avoid storing the data in an extra field is by extracting Line53-63 in operation.py to a function. And we call this new function on demand.

[the remark this answers is not in the document]
Comment (kazukiosawa) [reply asdfghjkl/operations/operation.py:29]: I see.
but both `_call_operations_in_forward()` and `_call_operatoins_in_backward()` receive the preprocessed `in_data`.
https://github.com/kazukiosawa/asdfghjkl/blob/126026ecdd01accfee9bcf328fce5f8cc2d7b420/asdfghjkl/core.py#L25-L31

so do we ready need this `_unit_wise-A`?

[the remark this answers is not in the document]
Comment (Cecilwang) [reply asdfghjkl/operations/operation.py]: Yes, the `in_data` of forward and backward are identical. But I want to get extended input. Now I create a new function called `extend_input` to do that and I will call it before `cov_unit_wise`.

[the remark this answers is not in the document]
Comment (kazukiosawa) [reply asdfghjkl/operations/operation.py]: now I get your point! yes, preparing `extend_input()` sounds a reasonable solution
